lib/virtual-fs/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+use std::any::Any;
+use std::fmt;
+use std::io::{self, Read, Seek, Write};
+use std::path::Path;
+use thiserror::Error;
+use tracing::debug;
+
+pub mod host_fs;
+//pub mod vfs_fs;
+pub mod mem_fs;
+
+pub trait FileSystem: Send + Sync + 'static {
+    fn read_dir(&self, path: &Path) -> Result<std::fs::ReadDir, FsError>;
+    fn create_dir(&self, path: &Path) -> Result<(), FsError>;
```

Comment:
Yeah, it could be, I'm not sure it's any better though, then users of `dyn FileSystem` will need to do the sharing themselves, which is a fine pattern, this way is kind of simpler for users of the API though

Reply:
Hmm, interesting, so the use case for Wasmer absolutely needs locking due to how `WasmerEnv` works and how we'd need  to make things work in WASI and emscripten... that's a good point though, if it's for use in Wasm, maybe it's worth it to move the locking to the API user as needed